File: AnalysysOCDemo/AnalysysOCDemo/AnalysysLogManager/AnalysysLogManager.m
//
//  AnalysysLogManager.m
//  AnalysysOCDemo
//
//  Created by xiao xu on 2020/7/27.
//  Copyright © 2020 xiao xu. All rights reserved.
//

#import "AnalysysLogManager.h"
#import "AnalysysLogVC.h"

#define ANSLogButtonWith 60
#define ANSLogButtonHeight 60

@interface AnalysysLogManager()

@property (nonatomic, strong)UIButton *floatBtn;

@end

@implementation AnalysysLogManager

+ (instancetype)sharedManager {
    static id singleInstance = nil;
    static dispatch_once_t onceToken ;
    dispatch_once(&onceToken, ^{
        if (!singleInstance) {
            singleInstance = [[self alloc] init] ;
        }
    });
    return singleInstance;
}

+ (void)showFloatingButton {
    UIButton *btn = [UIButton buttonWithType:UIButtonTypeCustom];
    [btn setTitle:@"日志" forState:UIControlStateNormal];
    btn.bounds = CGRectMake(0, 0, ANSLogButtonWith, ANSLogButtonHeight);
    btn.frame = CGRectMake(0, [UIScreen mainScreen].bounds.size.height/2.0 - ANSLogButtonWith/2.0, ANSLogButtonWith, ANSLogButtonHeight);
    btn.backgroundColor = [UIColor colorWithHexString:@"#686BF4"];
    [btn addTarget:self action:@selector(showLogVC) forControlEvents:UIControlEventTouchUpInside];
    btn.layer.cornerRadius = ANSLogButtonWith/2.0;
    
    UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(pan:)];
    [btn addGestureRecognizer:pan];
    
    [[UIApplication sharedApplication].keyWindow addSubview:btn];
    [AnalysysLogManager sharedManager].floatBtn = btn;
}

+ (void)showLogVC {
    [[[[UIApplication sharedApplication] delegate] window] endEditing:YES];
    
    AnalysysLogVC *analysysLogVC = [[AnalysysLogVC alloc] init];
    analysysLogVC.callBackBlock = ^{
        [AnalysysLogManager sharedManager].floatBtn.hidden = NO;
    };
    UINavigationController *nav = [[UINavigationController alloc] initWithRootViewController:analysysLogVC];
    nav.modalPresentationStyle = UIModalPresentationFullScreen;
    [[self getTopViewController] presentViewController:nav animated:true completion:^{
        [AnalysysLogManager sharedManager].floatBtn.hidden = YES;
    }];
}

+ (void)pan:(UIPanGestureRecognizer *)pan {
    UIButton *floatBtn = [AnalysysLogManager sharedManager].floatBtn;
    //获取偏移量
    // 返回的是相对于最原始的手指的偏移量
    CGPoint transP = [pan translationInView:floatBtn];
    
    // 移动图片控件
    floatBtn.transform = CGAffineTransformTranslate(floatBtn.transform, transP.x, transP.y);
    
    CGFloat btnWidth = floatBtn.frame.size.width;
    CGFloat btnHight = floatBtn.frame.size.height;
    CGFloat screenWidth = [UIScreen mainScreen].bounds.size.width;
    CGFloat screenHeight = [UIScreen mainScreen].bounds.size.height;
    if (floatBtn.frame.origin.x < 0) {
        floatBtn.frame = CGRectMake(0, floatBtn.frame.origin.y, ANSLogButtonWith, ANSLogButtonHeight);
    } else if (floatBtn.frame.origin.x > screenWidth - btnWidth) {
        floatBtn.frame = CGRectMake(screenWidth - btnWidth, floatBtn.frame.origin.y, ANSLogButtonWith, ANSLogButtonHeight);
    } else if (floatBtn.frame.origin.y < 0) {
        floatBtn.frame = CGRectMake(floatBtn.frame.origin.x, 0, ANSLogButtonWith, ANSLogButtonHeight);
    } else if (floatBtn.frame.origin.y > screenHeight - btnHight) {
        floatBtn.frame = CGRectMake(floatBtn.frame.origin.x, screenHeight - btnHight, ANSLogButtonWith, ANSLogButtonHeight);
    }
    
    // 复位,表示相对上一次
    [pan setTranslation:CGPointZero inView:floatBtn];
}

+ (UIViewController *)rootViewController {
    UIWindow* window = [[[UIApplication sharedApplication] delegate] window];
    NSAssert(window, @"The window is empty");
    return window.rootViewController;
}

+ (UIViewController *)getTopViewController {
    UIViewController* currentViewController = [self rootViewController];
    BOOL runLoopFind = YES;
    while (runLoopFind) {
        if (currentViewController.presentedViewController) {
            currentViewController = currentViewController.presentedViewController;
        } else {
            if ([currentViewController isKindOfClass:[UINavigationController class]]) {
                currentViewController = ((UINavigationController *)currentViewController).visibleViewController;
            } else if ([currentViewController isKindOfClass:[UITabBarController class]]) {
                currentViewController = ((UITabBarController* )currentViewController).selectedViewController;
            } else {
                break;
            }
        }
    }
    return currentViewController;
}


@end

File: AnalysysSwiftDemo/AnalysysSwiftDemo/AnalysysLogManager/AnalysysLogManager.swift
//
//  AnalysysLogManager.swift
//  AnalysysSwiftDemo
//
//  Created by xiao xu on 2020/7/28.
//  Copyright © 2020 xiao xu. All rights reserved.
//

import UIKit

class AnalysysLogManager: NSObject {
    
    weak var button: UIButton?
    
    static let sharedInstance: AnalysysLogManager = {
        let instance = AnalysysLogManager()
        // setup code
        return instance
    }()
    
    func createSuspendButton() -> Void {
        
        let btn: UIButton = UIButton(type: .custom)
        btn.setTitle("日志", for: UIControl.State.normal)
        btn.bounds = CGRect(x: 0, y: 0, width: 60, height: 60)
        btn.center = CGPoint(x: UIScreen.main.bounds.size.width/2, y: UIScreen.main.bounds.size.height/2)
        btn.backgroundColor = UIColor.colorWithHexString(hexadecimal: "#20a0ff")
        btn.addTarget(self, action: #selector(showLogVC), for: UIControl.Event.touchUpInside)
        btn.layer.cornerRadius = 30.0
        
        let pan: UIPanGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(pan(p:)))
        btn.addGestureRecognizer(pan)
        
        UIApplication.shared.keyWindow?.addSubview(btn)
        self.button = btn
        
    }
    
    
    @objc func showLogVC() -> Void {
        
        let analysysLogVC: AnalysysLogVC = AnalysysLogVC()
        let nav: UINavigationController = UINavigationController(rootViewController: analysysLogVC)
        self.ans_findVisibleViewController()?.present(nav, animated: true) {
            self.button?.isEnabled = false
        }
        
    }
    
    @objc func pan(p: UIPanGestureRecognizer) -> Void {
        
        let transform: CGAffineTransform = self.button!.transform
        let transP: CGPoint = p.translation(in: self.button)
        self.button?.transform = transform.translatedBy(x: transP.x, y: transP.y)
        p.setTranslation(CGPoint(x: 0, y: 0), in: self.button)
    }
    
    func onEventDataReceived(_ eventData: Any!) {
        if eventData != nil {
            AnalysysLogData.sharedSingleton.logData?.insert(eventData as Any, at: 0)
        }
    }
    
    func ans_rootViewController() -> UIViewController {
        
        let win: UIWindow = (UIApplication.shared.delegate?.window)!!
        return win.rootViewController!
    }
    
    func ans_findVisibleViewController() -> UIViewController? {
        
        var currentViewController: UIViewController = self.ans_rootViewController()
        let runLoopFind: Bool = true
        while runLoopFind {
            if (currentViewController.presentedViewController != nil) {
                currentViewController = currentViewController.presentedViewController!
            } else {
                if currentViewController.isKind(of: UINavigationController.self) {
                    currentViewController = (currentViewController as! UINavigationController).visibleViewController!
                } else if currentViewController.isKind(of: UITabBarController.self) {
                    currentViewController = (currentViewController as! UITabBarController).selectedViewController!
                } else {
                    break
                }
            }
        }
        return currentViewController
    }
}
